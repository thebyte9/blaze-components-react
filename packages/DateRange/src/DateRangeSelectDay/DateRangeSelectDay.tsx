// @ts-nocheck
import Input from "@blaze-react/input";
import withUtils from "@blaze-react/utils";
import React, { useEffect, useRef, useState } from "react";
import { MONTHS, NEXT, PREVIOUS, SEPARATOR } from "../constants";
import { IDateRangeProps } from "../interfaces";

import { DateUtils } from "../utils";

const DateRange: React.SFC<IDateRangeProps> = ({
  onChange,
  type,
  utils: { classNames }
}) => {
  const [year, setYear] = useState<string | number>(new Date().getFullYear());
  const [month, setMonth] = useState<string | number>(new Date().getMonth());
  const [date, setDate] = useState<string | null>(new Date().getDate());
  const [selectedDate, setSelectedDate] = useState<string>("");
  const [calendarStatus, setCalendarStatus] = useState<boolean>(false);
  const wrapperRef = useRef(null);

  const changeValue = (DD: string, MM: string, YY: string) => {
    const pickedDate = DateUtils.formatDate(DD, MM, YY);

    setSelectedDate(pickedDate);
    onChange({ [type.toLowerCase()]: pickedDate });
  };

  const handleCloseCalendar = (event: any) => {
    if (wrapperRef) {
      const { current } = wrapperRef;
      if (current && !current.contains(event.target)) {
        setCalendarStatus(false);
      }
    }
  };

  useEffect(() => {
    document.addEventListener("mousedown", handleCloseCalendar);
    return () => {
      document.removeEventListener("mousedown", handleCloseCalendar);
    };
  }, [wrapperRef]);

  const changeDate = (event: any) => {
    const { target: { dataset: { currentdate } = {} } = {} } = event;
    const [newDate, newMonth, newYear] = currentdate.split(SEPARATOR);

    const { DD, MM, YY } = DateUtils.dateToNumber(newDate, newMonth, newYear);
    changeValue(DD, MM + 1, YY);

    setYear(YY);
    setMonth(MM);
    setDate(DateUtils.formatDate(DD, MM, YY));
  };

  const handlePreviousMonth = () => {
    const { prevYear, prevMonth } = DateUtils.getPrevMonth(month, year);

    setYear(prevYear);
    setMonth(prevMonth);
  };

  const handleNextMonth = () => {
    const { nextYear, nextMonth } = DateUtils.getNextMonth(month, year);

    setYear(nextYear);
    setMonth(nextMonth);
  };

  const getDayClassName = (dateToCheck: any): string => {
    const pickedDate = DateUtils.formatDate(
      dateToCheck.getDate(),
      dateToCheck.getMonth(),
      dateToCheck.getFullYear()
    );

    return classNames({
      outDate: dateToCheck.getMonth() !== month,
      pointDate: pickedDate === date,
      today: DateUtils.isToday(dateToCheck) && pickedDate !== date
    });
  };

  const resetDate = () => {
    const newYear = DateUtils.currentDate.getFullYear();
    const newMonth = DateUtils.currentDate.getMonth();
    const newDate = DateUtils.currentDate.getDate();

    changeValue(newDate, newMonth + 1, newYear);

    setYear(newYear);
    setMonth(newMonth);
    setDate(newDate);
  };

  const handleInputChange = ({ value }) => {
    const [newDate, newMonth, newYear] = value.split(SEPARATOR);

    if (DateUtils.isInvalidDate(newDate, newMonth, newYear)) {
      return;
    }

    const { DD, MM, YY } = DateUtils.dateToNumber(newDate, newMonth, newYear);

    changeValue(DD, MM, YY);
    setYear(YY);
    setMonth(MM - 1);
    setDate(DateUtils.formatDate(DD, MM - 1, YY));
  };

  const handleOnClick = () => setCalendarStatus(true);

  return (
    <div className="calendar-container" ref={wrapperRef}>
      <div className="calendar-input" onClick={handleOnClick}>
        <span>{type}</span>
        <Input
          placeholder="dd/mm/yy"
          value={selectedDate}
          className="range-date"
          onChange={handleInputChange}
        />
      </div>
      {calendarStatus && (
        <div className="calendar" for={type}>
          <div className="header">
            <span className="prevMonth" onClick={handlePreviousMonth}>
              {PREVIOUS}
            </span>
            <div className="monthAndYear">
              <span onClick={resetDate}>
                {MONTHS[month]} {year}
              </span>
            </div>
            <span className="nextMonth" onClick={handleNextMonth}>
              {NEXT}
            </span>
          </div>
          <div className="calendar__content">
            <ul className="calendar__content__list">
              {DateUtils.getNameOfDays()}
            </ul>
            <ul className="calendar__content__days">
              {DateUtils.getListOfDays(year, month).map(
                ({
                  currentYearToString,
                  currentMonthToString,
                  currentDateToString,
                  currentDate
                }) => (
                  <li
                    className={`calendar__content__days__item day ${getDayClassName(
                      currentDate
                    )}`}
                    data-currentdate={DateUtils.formatDate(
                      currentDateToString,
                      currentMonthToString,
                      currentYearToString
                    )}
                    onClick={changeDate}
                  >
                    {currentDate.getDate()}
                  </li>
                )
              )}
            </ul>
          </div>
        </div>
      )}
    </div>
  );
};

DateRange.defaultProps = {
  onChange: () => void 0,
  type: "selectedDate"
};
export default withUtils(DateRange);
