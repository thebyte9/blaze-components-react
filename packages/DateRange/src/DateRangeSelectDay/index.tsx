import DateRangeSelectDay from "./DateRangeSelectDay";

export default DateRangeSelectDay;
