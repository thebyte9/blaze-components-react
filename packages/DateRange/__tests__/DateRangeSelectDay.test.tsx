import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import DateRangeSelectDay from '../src/DateRangeSelectDay';
import React from 'react';

describe('DateRange component', () => {
  test('should be defined and renders correctly (snapshot)', () => {
    const { asFragment } = render(<DateRangeSelectDay onChange={jest.fn} type="custom" />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should select a value', () => {
    render(<DateRangeSelectDay onChange={jest.fn} type="custom" />);
    const input = screen.getByTestId('custom');
    fireEvent.change(input, { target: { value: '11/12/2021' } });
  });

  test('should handle opening and closing the calendar', () => {
    render(<DateRangeSelectDay onChange={jest.fn} type="custom" />);

    fireEvent.click(screen.getByTestId('custom'));
    expect(screen.getByTestId('calendar-container')).toBeInTheDocument();

    fireEvent.mouseDown(screen.getByTestId('custom'));
    expect(screen.getByTestId('calendar-container')).toBeInTheDocument();
  });

  test('should handle changing dates', () => {
    render(<DateRangeSelectDay onChange={jest.fn} type="custom" />);

    fireEvent.change(screen.getByTestId('custom'), { target: { value: '12/11/2021' } });
    fireEvent.click(screen.getByTestId('custom'));
    fireEvent.mouseDown(screen.getByTestId('custom'));
  });

  test('should handle prev month', () => {
    render(<DateRangeSelectDay onChange={jest.fn} type="custom" />);

    fireEvent.change(screen.getByTestId('custom'), { target: { value: '12/11/2021' } });
    fireEvent.click(screen.getByTestId('custom'));
    fireEvent.click(screen.getByTestId('calendar-header-prev'));
  });

  test('should handle invalid dates', () => {
    render(<DateRangeSelectDay onChange={jest.fn} type="custom" />);

    fireEvent.change(screen.getByTestId('custom'), { target: { value: '1' } });
  });

  test('should handle next month', () => {
    render(<DateRangeSelectDay onChange={jest.fn} type="custom" />);

    fireEvent.change(screen.getByTestId('custom'), { target: { value: '12/11/2021' } });
    fireEvent.click(screen.getByTestId('custom'));
    fireEvent.click(screen.getByTestId('calendar-header-next-month'));
  });

  test('should reset header', () => {
    render(<DateRangeSelectDay onChange={jest.fn} type="custom" />);

    fireEvent.change(screen.getByTestId('custom'), { target: { value: '12/11/2021' } });
    fireEvent.click(screen.getByTestId('custom'));
    fireEvent.click(screen.getByTestId('calendar-reset-date'));
  });
});
