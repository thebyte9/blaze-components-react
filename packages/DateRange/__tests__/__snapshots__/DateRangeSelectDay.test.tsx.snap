// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRange component should be defined and renders correctly (snapshot) 1`] = `
<DocumentFragment>
  <div
    class="calendar-container"
  >
    <div
      class="calendar-input"
    >
      <span
        class="calendar-input__label"
      >
        selectedDate
      </span>
      <div
        class="form-field form-field--input  "
      >
        <label
          class=""
        />
        <input
          class="range-date"
          data-testid="input"
          placeholder="dd/mm/yy"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
